devito/data/data.py

```diff
@@ -132,16 +142,82 @@ def _global(self, glb_idx, decomposition):
         ret._is_distributed = any(i is not None for i in decomposition)
         return ret
 
+    def _check_idx(func):
+        """Check if __getitem__/__setitem__ may require communication across MPI ranks."""
+        @wraps(func)
+        def wrapper(data, *args, **kwargs):
+            glb_idx = args[0]
+            if len(args) > 1 and isinstance(args[1], Data) \
+                    and args[1]._is_mpi_distributed:
+                comm_type = PARALLEL
+            elif data._is_mpi_distributed:
+                for i in as_tuple(glb_idx):
+                    if isinstance(i, slice) and i.step is not None and i.step < 0:
+                        comm_type = PARALLEL
+                        break
+                    else:
+                        comm_type = SERIAL
+            else:
+                comm_type = SERIAL
+            kwargs['comm_type'] = comm_type
+            return func(data, *args, **kwargs)
+        return wrapper
+
     @property
     def _is_mpi_distributed(self):
         return self._is_distributed and configuration['mpi']
 
     def __repr__(self):
         return super(Data, self._local).__repr__()
 
-    def __getitem__(self, glb_idx):
-        loc_idx = self._convert_index(glb_idx)
-        if loc_idx is NONLOCAL:
+    @_check_idx
+    def __getitem__(self, glb_idx, comm_type):
+        loc_idx = self._index_glb_to_loc(glb_idx)
+        if comm_type is PARALLEL:
+            # Retrieve the pertinent local data prior to mpi send/receive operations
+            loc_data_idx = local_data_idx(loc_idx)
+            local_val = super(Data, self).__getitem__(loc_data_idx)
+
+            comm = self._distributor.comm
+            rank = comm.Get_rank()
+
+            owners, send, global_si, local_si = \
+                mpi_index_maps(loc_idx, local_val.shape, self._distributor.topology,
+                               self._distributor.all_coords, comm)
+
+            retval = Data(local_val.shape, local_val.dtype.type,
+                          decomposition=local_val._decomposition,
```

Comment:
Could you clarify with a comment why are you passing in this ``local_val._decomposition``  ?

Reply:
Numpy `__getitem__` loops though each element. So after the 'parallel' `__getitem__` there'll be a load of serial `__getitem__` calls in which we'll still need to work out the `loc_idx` which needs the decomposition.